src/Type/PropertyCast.php

```diff
@@ -0,0 +1,48 @@
+<?php
+
+namespace Rexlabs\DataTransferObject\Type;
+
+use const Rexlabs\DataTransferObject\NONE;
+
+interface PropertyCast
+{
+    /**
+     * @param string $type
+     *
+     * @return bool
+     */
+    public function canCastType(string $type): bool;
+
+    /**
+     * @param mixed $value
+     *
+     * @return bool
+     */
+    public function shouldCastValue($value): bool;
+
+    /**
+     * @param mixed $property
+     *
+     * @return bool
+     */
+    public function shouldMapToData($property): bool;
+
+    /**
+     * @param string $name
+     * @param mixed $data
+     * @param string $type
+     * @param int $flags
+     *
+     * @return mixed
+     */
+    public function castToType(string $name, $data, string $type, int $flags = NONE);
+
+    /**
+     * @param string $name
+     * @param mixed $property
+     * @param int $flags
+     *
+     * @return mixed
+     */
+    public function toData(string $name, $property, int $flags = NONE);
+}
```

Comment:
You've tried to document the purpose of each of these methods in the documentation but not inline here - any reason?

I love the implementation but am really struggling with the variation in the method names.

Do we need shouldCastValue / shouldMapToData - if we shouldn't cast the value can't castToType just return the value we passed in?

How do you feel about this.

```php
interface PropertyCast
{
    /**
     * @param string $type
     *
     * @return bool
     */
    public function canCastType(string $type){
        return $type === BlahBlah::class;
    }

    /**
     * @param string $name
     * @param mixed $data
     * @param string $type
     * @param int $flags
     *
     * @return mixed
     */
    public function toType(string $name, $data, string $type, int $flags = NONE){
        //if we shouldn't cast for whatever reason, just return the data
        if( !is_array($data) ){
            return $data;
        }

        return BlahBlah::make($data);
    }

    /**
     * @param string $name
     * @param mixed $property
     * @param int $flags
     *
     * @return mixed
     */
    public function toData(string $name, $property, int $flags = NONE){
        return $property ? $property->toArray() : $property;
    }
}
```

Reply:
> You've tried to document the purpose of each of these methods in the documentation but not inline here - any reason?

It's just the order of the PRs - I wrote those docs much later. I'll port whatever docs end up working back into the class later.

> I love the implementation but am really struggling with the variation in the method names.

Me too, I feel like it's not obvious enough how to write a PropertyCast.

> Do we need shouldCastValue / shouldMapToData - if we shouldn't cast the value can't castToType just return the value we passed in?

That does sound good. The main benefit of the current implementation is that it avoids cast collisions where, cast 1 turned an array into an object and cast 2 also expected an array and threw. I think it's possible to make the dispatcher smarter so the casts don't have to be like you're suggesting. I'll make an attempt now.

> How do you feel about this.

I like it, I'm just not sure I can make that work. I'm taking another look at the internals now but since most properties are union types it gets complicated.

Reply:
Ok I changed how it worked a bit and the above suggestion works really well. It even enabled a new use case where we cast all items of an array to DTOs and then the array of DTOs gets cast to a collection type. So we'll be able to write properties with types like `Collection|UserData[]` and it'll cast automatically.